lib/bolt/cli.rb

```diff
@@ -483,21 +494,39 @@ def outputter
       @outputter ||= Bolt::Outputter.for_format(@config[:format])
     end
 
-    def run_task(name, nodes, args, &block)
-      parse_error = nil
-      Puppet.initialize_settings
+    def in_bolt_compiler(opts = [])
+      Puppet.initialize_settings(opts)
       Puppet::Pal.in_tmp_environment('bolt', modulepath: [BOLTLIB_PATH] + @config[:modulepath], facts: {}) do |pal|
         pal.with_script_compiler do |compiler|
-          begin
-            return compiler.call_function('run_task', name, nodes, args, &block)
-          rescue Puppet::ParseError => e
-            # we assume that the Puppet::ParseError is due to a problem with
-            # the task and/or its parameters, but since the with_script_compiler
-            # method rescues these errors and raises different ones instead,
-            # we save it in a local variable and raise it as Bolt::CLIError
-            # outside of the block
-            parse_error = e
-          end
+          yield compiler
+        end
+      end
+    end
+
+    def list_tasks
+      in_bolt_compiler do |compiler|
+        tasks = compiler.list_tasks
+        tasks.map(&:name).sort.map do |task_name|
+          task_sig = compiler.task_signature(task_name)
+          [task_name, task_sig.task.description]
+        end
+      end
+    rescue Puppet::Error
+      raise Bolt::CLIError, "Failure while reading task metadata"
```

Comment:
Updated to catch Puppet exceptions so we don't print them twice (and don't print a stack trace).